### PR TITLE
BPH Offline Monitoring

### PR DESCRIPTION
**Pull request by anigamova**
tnp  added. new paths and variables

### CODE REVIEW

**Comment by mtosi** [comment]
@anigamova, please make this PR for master (92x)

File: DQMOffline/Trigger/plugins/BPHMonitor.cc [hunk 1]
```diff
@@ -146,55 +314,334 @@ void BPHMonitor::bookHistograms(DQMStore::IBooker     & ibooker,
...
-      mud0_.numerator ->Fill(d0);
-      muz0_.numerator ->Fill(z0);
+  std::vector<reco::Muon> tagMuons;
+  std::vector<reco::Muon> probeMuons;
+  for ( auto const & m : *muoHandle ) {
+  if ( muoSelection_tag( m ) ) tagMuons.push_back(m);
+  }
+  for (int i = 0; i<int(tagMuons.size());i++){
+    for ( auto const & m : *muoHandle ) { 
+      if ((tagMuons[i].pt() - m.pt())<=0.01)continue;//not the same
```
Comment (mtosi) [inline DQMOffline/Trigger/plugins/BPHMonitor.cc]: can we have this "magic" numbers set as global variables
and not having them hard coded, please ?

File: DQMOffline/Trigger/plugins/BPHMonitor.cc [hunk 2]
```diff
@@ -146,55 +314,334 @@ void BPHMonitor::bookHistograms(DQMStore::IBooker     & ibooker,
...
-      muz0_.numerator ->Fill(z0);
+  std::vector<reco::Muon> tagMuons;
+  std::vector<reco::Muon> probeMuons;
+  for ( auto const & m : *muoHandle ) {
+  if ( muoSelection_tag( m ) ) tagMuons.push_back(m);
+  }
+  for (int i = 0; i<int(tagMuons.size());i++){
+    for ( auto const & m : *muoHandle ) { 
+      if ((tagMuons[i].pt() - m.pt())<=0.01)continue;//not the same  
+      if ((tagMuons[i].p4()+m.p4()).M() >2.596&& (tagMuons[i].p4()+m.p4()).M() <3.596){//near to J/psi mass
```
Comment (mtosi) [inline DQMOffline/Trigger/plugins/BPHMonitor.cc]: above all these !

File: DQMOffline/Trigger/plugins/BPHMonitor.cc [hunk 3]
```diff
@@ -146,55 +314,334 @@ void BPHMonitor::bookHistograms(DQMStore::IBooker     & ibooker,
...
     }
+      
+  }
+    
+
+  }  
+  else{
+    if (trOrMu_){
+      edm::Handle<reco::TrackCollection> trHandle;
+      iEvent.getByToken( trToken_, trHandle );
```
Comment (mtosi) [inline DQMOffline/Trigger/plugins/BPHMonitor.cc]: please, add the check isValid()

File: DQMOffline/Trigger/plugins/BPHMonitor.cc [hunk 4]
```diff
@@ -146,55 +314,334 @@ void BPHMonitor::bookHistograms(DQMStore::IBooker     & ibooker,
...
+          j_tks.push_back(mu2TT);
+          if (j_tks.size()!=2) continue;
+       
+          KalmanVertexFitter jkvf;
+          TransientVertex jtv = jkvf.vertex(j_tks);
+          if (!jtv.isValid()) continue;
+        
+          reco::Vertex jpsivertex = jtv;
+          float dimuonCL = 0;
+          if( (jpsivertex.chi2()>=0.0) && (jpsivertex.ndof()>0) )
```
Comment (mtosi) [inline DQMOffline/Trigger/plugins/BPHMonitor.cc]: these numbers as well

File: DQMOffline/Trigger/plugins/BPHMonitor.cc [hunk 5]
```diff
@@ -146,55 +314,334 @@ void BPHMonitor::bookHistograms(DQMStore::IBooker     & ibooker,
...
+  else{
+  if (den_genTriggerEventFlag_->on() && ! den_genTriggerEventFlag_->accept( iEvent, iSetup) ) return;
+  for (auto const & m : *muoHandle ) {
+  for (auto const & m1 : *muoHandle ) {
+        
+        if (m1.pt() == m.pt())continue;
+      if(!muoSelection_ref(m))continue;   
+      if(!muoSelection_ref(m1))continue;   
+      switch(nofset_){
+      case 1: tnp_=1;
```
Comment (mtosi) [inline DQMOffline/Trigger/plugins/BPHMonitor.cc]: can you add few comments for helping the not expert to get the different cases, please ?

File: DQMOffline/Trigger/plugins/BPHMonitor.cc [hunk 6]
```diff
@@ -13,16 +13,32 @@ BPHMonitor::BPHMonitor( const edm::ParameterSet& iConfig ) :
   folderName_             ( iConfig.getParameter<std::string>("FolderName") )
   , muoToken_             ( mayConsume<reco::MuonCollection>       (iConfig.getParameter<edm::InputTag>("muons")     ) )  
   , bsToken_              ( mayConsume<reco::BeamSpot>             (iConfig.getParameter<edm::InputTag>("beamSpot")))
+  , trToken_              ( mayConsume<reco::TrackCollection>             (iConfig.getParameter<edm::InputTag>("tracks")))
+  , phToken_              ( mayConsume<reco::PhotonCollection>             (iConfig.getParameter<edm::InputTag>("photons")))
   , phi_binning_          ( getHistoPSet   (iConfig.getParameter<edm::ParameterSet>("histoPSet").getParameter<edm::ParameterSet>   ("phiPSet")    ) )
   , pt_binning_          ( getHistoPSet   (iConfig.getParameter<edm::ParameterSet>("histoPSet").getParameter<edm::ParameterSet>   ("ptPSet")    ) )
```
Comment (mtosi) [inline DQMOffline/Trigger/plugins/BPHMonitor.cc]: I think we should have variable binning

File: DQMOffline/Trigger/plugins/BPHMonitor.cc [hunk 7]
```diff
@@ -13,16 +13,32 @@ BPHMonitor::BPHMonitor( const edm::ParameterSet& iConfig ) :
...
   , eta_binning_          ( getHistoPSet   (iConfig.getParameter<edm::ParameterSet>("histoPSet").getParameter<edm::ParameterSet>   ("etaPSet")    ) )
```
Comment (mtosi) [inline DQMOffline/Trigger/plugins/BPHMonitor.cc]: I think we should have variable binning

File: DQMOffline/Trigger/plugins/BPHMonitor.cc [hunk 8]
```diff
@@ -13,16 +13,32 @@ BPHMonitor::BPHMonitor( const edm::ParameterSet& iConfig ) :
...
   , d0_binning_           ( getHistoLSPSet (iConfig.getParameter<edm::ParameterSet>("histoPSet").getParameter<edm::ParameterSet>   ("d0PSet")     ) )
```
Comment (mtosi) [inline DQMOffline/Trigger/plugins/BPHMonitor.cc]: I think we should have variable binning

File: DQMOffline/Trigger/python/BPHMonitor_cfi.py [hunk 9]
```diff
@@ -30,25 +31,65 @@
...
 hltBPHmonitoring.muons     = cms.InputTag("muons") # 
+hltBPHmonitoring.photons     = cms.InputTag("photons") # 
 
 hltBPHmonitoring.numGenericTriggerEventPSet.andOr         = cms.bool( False )
 hltBPHmonitoring.numGenericTriggerEventPSet.dbLabel       = cms.string("BPHDQMTrigger") # it does not exist yet, we should consider the possibility of using the DB, but as it is now it will need a label per path !
 hltBPHmonitoring.numGenericTriggerEventPSet.andOrHlt      = cms.bool(True)# True:=OR; False:=AND
+hltBPHmonitoring.numGenericTriggerEventPSet.andOrL1      = cms.bool(True)# True:=OR; False:=AND
 hltBPHmonitoring.numGenericTriggerEventPSet.hltInputTag   = cms.InputTag( "TriggerResults::HLT" )
-hltBPHmonitoring.numGenericTriggerEventPSet.hltPaths      = cms.vstring("HLT_Dimuon10_Jpsi_Barrel") # HLT_ZeroBias_v*
+hltBPHmonitoring.numGenericTriggerEventPSet.hltPaths      = cms.vstring("HLT_Mu7p5_TkX_Jpsi") # HLT_ZeroBias_v*
```
Comment (mtosi) [inline DQMOffline/Trigger/python/BPHMonitor_cfi.py]: is this available ?

File: DQMOffline/Trigger/python/BPHMonitor_cfi.py [hunk 10]
```diff
@@ -30,25 +31,65 @@
...
-hltBPHmonitoring.numGenericTriggerEventPSet.verbosityLevel = cms.uint32(1)
+hltBPHmonitoring.numGenericTriggerEventPSet.errorReplyL1 = cms.bool( True )
+hltBPHmonitoring.numGenericTriggerEventPSet.l1BeforeMask = cms.bool( True )
+hltBPHmonitoring.numGenericTriggerEventPSet.verbosityLevel = cms.uint32(0)
 
 hltBPHmonitoring.denGenericTriggerEventPSet.andOr         = cms.bool( False )
 hltBPHmonitoring.denGenericTriggerEventPSet.dcsInputTag   = cms.InputTag( "scalersRawToDigi" )
 hltBPHmonitoring.denGenericTriggerEventPSet.hltPaths  = cms.vstring( "HLT_Dimuon6_Jpsi_NoVertexing" )#reference
-hltBPHmonitoring.denGenericTriggerEventPSet.dcsPartitions = cms.vint32 ( 0,1,2,3,5,6,7,8,9,12,13,14,15,16,17,20,22,24, 25, 26, 27, 28, 29 ) # 24-27: strip, 28-29: pixel, we should add all other detectors !TODO
+hltBPHmonitoring.denGenericTriggerEventPSet.l1Algorithms      = cms.vstring("DM0_SQ") # HLT_ZeroBias_v*
```
Comment (mtosi) [inline DQMOffline/Trigger/python/BPHMonitor_cfi.py]: did you test it ?

File: DQMOffline/Trigger/python/BPHMonitor_cfi.py [hunk 11]
```diff
@@ -30,25 +31,65 @@
...
 
 hltBPHmonitoring.numGenericTriggerEventPSet.andOr         = cms.bool( False )
 hltBPHmonitoring.numGenericTriggerEventPSet.dbLabel       = cms.string("BPHDQMTrigger") # it does not exist yet, we should consider the possibility of using the DB, but as it is now it will need a label per path !
 hltBPHmonitoring.numGenericTriggerEventPSet.andOrHlt      = cms.bool(True)# True:=OR; False:=AND
+hltBPHmonitoring.numGenericTriggerEventPSet.andOrL1      = cms.bool(True)# True:=OR; False:=AND
 hltBPHmonitoring.numGenericTriggerEventPSet.hltInputTag   = cms.InputTag( "TriggerResults::HLT" )
-hltBPHmonitoring.numGenericTriggerEventPSet.hltPaths      = cms.vstring("HLT_Dimuon10_Jpsi_Barrel") # HLT_ZeroBias_v*
+hltBPHmonitoring.numGenericTriggerEventPSet.hltPaths      = cms.vstring("HLT_Mu7p5_TkX_Jpsi") # HLT_ZeroBias_v*
+hltBPHmonitoring.numGenericTriggerEventPSet.l1Algorithms      = cms.vstring("DM0_SQ") # HLT_ZeroBias_v*
 hltBPHmonitoring.numGenericTriggerEventPSet.hltDBKey      = cms.string("diMu10")
```
Comment (mtosi) [inline DQMOffline/Trigger/python/BPHMonitor_cfi.py]: are you going to use a payload ?

**Comment by mtosi** [comment]
I'm confused by the list of triggers you are monitoring

where are 
HLT_Trimuon2_Upsilon5_Muon_v1
HLT_DoubleMu4_3_Bs_v8
?

**Comment by mtosi** [comment]
HLT_Dimuon20_Jpsi_Barrel_Seagulls_v1
HLT_Dimuon10_PsiPrime_Barrel_Seagulls_v1
HLT_Dimuon10_Upsilon_Barrel_Seagulls_v1
HLT_Dimuon14_Phi_Barrel_Seagulls_v1

**Comment by mtosi** [comment]
HLT_DoubleMu4_JpsiTrk_Displaced_v7
HLT_DoubleMu4_LowMassNonResonantTrk_Displaced_v7
HLT_DoubleMu4_PsiPrimeTrk_Displaced_v7
# HLT_Mu7p5_L2Mu2_Jpsi_v5
HLT_Mu7p5_Track2_Jpsi_v4
HLT_Mu7p5_Track3p5_Jpsi_v4
HLT_Mu7p5_Track7_Jpsi_v4
# HLT_Dimuon0_Jpsi_L1_NoOS_v1
# HLT_Dimuon0_Jpsi_NoVertexing_NoOS_v1
# HLT_Dimuon0_Jpsi_v1
# HLT_Dimuon0_Jpsi_NoVertexing_v1
# HLT_Dimuon0_Jpsi_L1_4R_0er1p5R_v1
# HLT_Dimuon0_Jpsi_NoVertexing_L1_4R_0er1p5R_v1
HLT_Dimuon0_LowMass_L1_0er1p5R_v1
# HLT_Dimuon0_LowMass_L1_0er1p5_v1
# HLT_Dimuon0_LowMass_v1
# HLT_Dimuon0_LowMass_L1_4_v1
HLT_Dimuon0_LowMass_L1_4R_v1
HLT_DoubleMu4_Jpsi_Displaced_v1
HLT_DoubleMu4_Jpsi_NoVertexing_v1
# HLT_Mu7p5_L2Mu2_Upsilon_v5
HLT_Mu7p5_Track2_Upsilon_v4
HLT_Mu7p5_Track3p5_Upsilon_v4
HLT_Mu7p5_Track7_Upsilon_v4
# HLT_Dimuon0_Upsilon_L1_4p5_v1
# HLT_Dimuon0_Upsilon_L1_5_v1
# HLT_Dimuon0_Upsilon_L1_4p5NoOS_v1
# HLT_Dimuon0_Upsilon_L1_4p5er2p0_v1
# HLT_Dimuon0_Upsilon_L1_4p5er2p0M_v1
# HLT_Dimuon0_Upsilon_NoVertexing_v1
# HLT_Dimuon0_Upsilon_L1_5M_v1
# HLT_DoubleMu20_7_Mass0to30_L1_DM4_v1
HLT_DoubleMu20_7_Mass0to30_L1_DM4EG_v1

**Comment by anigamova** [comment]
Hi, Mia! 
I thought Alessio has discussed the total strategy with you.
Here you can find the list of paths  that we as far as I understand agreed to monitor
https://docs.google.com/spreadsheets/d/1VkVioMr63WIEV-KWORUQpjmGdSUVsH3cx0VoREDO1VM/edit#gid=0

**Comment by anigamova** [comment]
Probably we should continue discussion with him.

File: DQMOffline/Trigger/plugins/BPHMonitor.cc (hunk 5 above)
Comment (anigamova) [inline DQMOffline/Trigger/plugins/BPHMonitor.cc]: These cases are for different sets of variables to monitor the efficiency. 
They are listed in this table, nofset variable(monitoring code) column
https://docs.google.com/spreadsheets/d/1VkVihttps://docs.google.com/spreadsheets/d/1VkVioMr63WIEV-KWORUQpjmGdSUVsH3cx0VoREDO1VM/edit#gid=0oMr63WIEV-KWORUQpjmGdSUVsH3cx0VoREDO1VM/edit#gid=0

**Comment by mtosi** [comment]
please, add few comments w//in the code
thanks
 mia

**Comment by mtosi** [comment]
@dmitrijus can we start the integration test, please ?

**Comment by fwyzard** [comment]
tracked by #19142

**Comment by anigamova** [comment]
@mtosi do we need this PR to be merged also? because I thought this one is useless since we have #18994 
Probably I should even close this PR

**Comment by mtosi** [comment]
if this is not needed any longer, please just close it

**Pull request closed without merging**